Gems/AWSCore/Code/Include/Framework/AWSApiRequestJob.h

```diff
@@ -153,16 +153,16 @@ namespace AWSCore
 
 /// An Az::Job that that executes a specific AWS request.
     template<class RequestTraits>
-    class AwsApiRequestJob
+    class AwsApiRequestJobBase
```

Comment:
Not a big fan of this fix, as it introduces a new layer of inheritance just for fixing a new compiler error.

I would suggest to try something like this by using forward declare to avoid incomplete class error:
```
template<class RequestTraits>
class AwsApiRequestJob
{
    ...
    class Function;
    ...
}

template<class RequestTraits>
class AwsApiRequestJob<RequestTraits>::Function 
    : public AwsApiRequestJob
{

}
```

Reply:
not sure, as all these code are copied from LY code base directly, probably we don't need to inherit AwsApiRequestJob for Function.
but that can be a different task to simplify Function, here I am incline to keep original design by fixing this error.

Reply:
one concern for moving Function outside, as you can see there is following static member function for AwsApiRequestJob class.
```
static AwsApiRequestJob<RequestTraits>* Create
```
It is open to o3de developer, if we move this function out of AwsApiRequestJob, it will break existing consumer code. (it will be a backward incompatible change?)

Reply:
I have updated the snippet to still have the `Create` function be a static member of the `AwsApiRequestJob<RequestTraits>` primary template, but I have moved the implementation outside of the `AwsApiRequestJob` template after the `Function<RequestJob>` template has been defined.

Reply:
cool, this approach looks good to me as well. The only minor is Function scope gets changed, but not really a big problem. (considering developer doesn't need to use it directly)

Reply:
From what I can tell, Function inheriting from AwsApiRequestJob is needed because that is the instance that is returned from the static Create function. The reason it is an inner class is because it was because it needs to be defined within the context of the 'Create' function. The reason why the create function is static and inside the AwsApiRequestJob is so the calling sites to this can do the following:

```
using PostProducerEventsRequestJob = AWSCore::ServiceRequestJob<PostProducerEventsRequest>;
```
and
```
 ServiceAPI::PostProducerEventsRequestJob* requestJob = ServiceAPI::PostProducerEventsRequestJob::Create(
            [this, requestId](ServiceAPI::PostProducerEventsRequestJob* successJob)
            {
                OnResponseReceived(successJob->parameters.data, successJob->result.events);

                AZ::TickBus::QueueFunction([requestId]()
                {
                    AWSMetricsNotificationBus::Broadcast(&AWSMetricsNotifications::OnSendMetricsSuccess, requestId);
                });
            },
            [this, requestId](ServiceAPI::PostProducerEventsRequestJob* failedJob)
            {
                OnResponseReceived(failedJob->parameters.data);

                AZStd::string errorMessage = failedJob->error.message;
                AZ::TickBus::QueueFunction([requestId, errorMessage]()
                {
                    AWSMetricsNotificationBus::Broadcast(&AWSMetricsNotifications::OnSendMetricsFailure, requestId, errorMessage);
                });
            });

        requestJob->parameters.data = AZStd::move(metricsQueue);
```

So the reason why the new level of inheritance was done for this case was so I minimize the impact of the clients of this class. If we pull Function out completely, then the caller sites would need an update (I did try this approach), and it started to affect more of the code base. 

Forward declaring won't work either because of the static create function i am trying to preserve to reduce the impact. The Function class needs to be defined before the Create function where it is instantiated.

Reply:
```
static AwsApiRequestJob* Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure = OnFailureFunction{}, IConfig* config = GetDefaultConfig())
```
for this static function itself, it doesn't need to know anything about Function when declare it. So can you move the definition out and put it after Function class definition?

Reply:
If the static function is moved out of the class, then it affects code that is using it, like the 'ServiceAPI::PostProducerEventsRequestJob' example above. It will only work if we go to all the places that use these classes and change the <classname>::Create(..) function, which is something I want to avoid for now

I'm currently looking into an alternate approach where we dont need s derived Function at all, but to describe the current solution in this PR:

1. Adding a layer of inheritance is just so that the original ::Create function can be used and no need to change other classes
2. The inherited Function class inside the new inherited class doesnt even have to be an inner class, we can pull it out as well
3. An alternative is to remove the Function derived class altogether and make updates to the OnSuccess and OnFailure of the original class to supported the intention of the Create function

Reply:
Sure no problem, the most native solution is to remove inheritance from function just copy base definition manually.

btw, when I say move static function definition out is like following
```
template<class RequestTraits>
class AwsApiRequestJob
{
    ...
    class Function;
    static AwsApiRequestJob* Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure = OnFailureFunction{}, IConfig* config = GetDefaultConfig());
}

// You are free to move this Function out of AwsApiRequestJob scope, if you do so, just remove forward declare above
// not a big problem
template<class RequestTraits>
class AwsApiRequestJob<RequestTraits>::Function 
    : public AwsApiRequestJob
{

}

template<class RequestTraits>
AwsApiRequestJob<RequestTraits>* AwsApiRequestJob<RequestTraits>::Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure, IConfig* config)
{
    return azcreate(Function, (onSuccess, onFailure, config), AZ::SystemAllocator);
}
```

so Create function still belongs to AwsApiRequestJob class, you won't need to change consumer code.

Reply:
I am testing alternative #3 right now. It looks like the only purpose for the internal Function derived classes was to queue onSuccess, on Failure, and DoCleanup events instead of doing the deletes automatically. 
So rather than doing that, i moved those queue events into the base class with a flag that controls that behavior instead. This will still preserve the consumer code

Reply:
> Sure no problem, the most native solution is to remove inheritance from Function class just copy base definition manually.
> 
> btw, when I say move static function definition out is like following
> 
> ```
> template<class RequestTraits>
> class AwsApiRequestJob
> {
>     ...
>     class Function;
>     static AwsApiRequestJob* Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure = OnFailureFunction{}, IConfig* config = GetDefaultConfig());
> }
> 
> // You are free to move this Function out of AwsApiRequestJob scope, if you do so, just remove forward declare above
> // not a big problem
> template<class RequestTraits>
> class AwsApiRequestJob<RequestTraits>::Function 
>     : public AwsApiRequestJob
> {
> 
> }
> 
> template<class RequestTraits>
> AwsApiRequestJob<RequestTraits>* AwsApiRequestJob<RequestTraits>::Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure, IConfig* config)
> {
>     return azcreate(Function, (onSuccess, onFailure, config), AZ::SystemAllocator);
> }
> ```
> 
> so Create function still belongs to AwsApiRequestJob class, you won't need to change consumer code.

I like this suggestion, I will try to get it to work

Reply:
> Sure no problem, the most native solution is to remove inheritance from Function class just copy base definition manually.
> 
> btw, when I say move static function definition out is like following
> 
> ```
> template<class RequestTraits>
> class AwsApiRequestJob
> {
>     ...
>     class Function;
>     static AwsApiRequestJob* Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure = OnFailureFunction{}, IConfig* config = GetDefaultConfig());
> }
> 
> // You are free to move this Function out of AwsApiRequestJob scope, if you do so, just remove forward declare above
> // not a big problem
> template<class RequestTraits>
> class AwsApiRequestJob<RequestTraits>::Function 
>     : public AwsApiRequestJob
> {
> 
> }
> 
> template<class RequestTraits>
> AwsApiRequestJob<RequestTraits>* AwsApiRequestJob<RequestTraits>::Create(OnSuccessFunction onSuccess, OnFailureFunction onFailure, IConfig* config)
> {
>     return azcreate(Function, (onSuccess, onFailure, config), AZ::SystemAllocator);
> }
> ```
> 
> so Create function still belongs to AwsApiRequestJob class, you won't need to change consumer code.

Yeah, this would've worked as well. But with #3 we can remove that unnecessary inheritance